File: Matlab/aitken.m

function aitken(f,x0,tolerance,nMax)
    fxi = f(x0);
    if fxi == 0
        disp(['A root was found: ', num2str(x0,12)])
    else
        matrix = ["counter","xi","fxi","error"];
        matrix = [matrix;[1,x0,fxi,2]];
        x1 = f(x0);
        error = abs(x1-x0);
        fxi = f(x1);
        matrix = [matrix;[2,x1,fxi,error]];
        x2 = f(x1);
        error = abs(x2-x1);
        fxi = f(x2);
        matrix = [matrix;[3,x2,fxi,error]];
        det = (x2-x1)-(x1-x0);
        counter = 4;
        while(fxi ~= 0 && error > tolerance && counter < nMax && det~=0)
            xi = x2 - ( (x2 - x1)^2 )/det;
            fxi = f(xi);
            det = (x2-x1)-(x1-x0);
            error = abs(xi-x2);
            matrix = [matrix;[counter,xi,fxi,error]];
            x0 = xi;
            x1 = f(x0);
            x2 = f(x1);
            counter = counter + 1;
        end
        
        disp(matrix)

        if fxi == 0 
            disp(['The root has been found and it is: ', num2str(xi,12)])
        elseif error <= tolerance
            disp(['An approximation has been found and is: ', num2str(xi,12)])
        elseif det == 0
            disp('Error during method execution')
        else 
            disp('The method fails with the maximum number of iterations given')
        end
        
    end
end

File: NumericSquadProject/numericApp/methods/Roots/aitken.py
import sympy as sm 

x = sm.symbols('x')

def aitken(f,x0,tol,n):
    f = sm.sympify(f)
    fxi = float(f.subs(x,x0))
    if fxi == 0:
        return "A root was found",[]
    else:
        matrix = []
        matrix.append([1,x0,fxi,1])
        x1 = float(f.subs(x,x0))
        error = abs(x1-x0)
        fxi = float(f.subs(x,x1))
        matrix.append([2,x1,fxi,error])
        x2 = float(f.subs(x,x1))
        error = abs(x2-x1)
        fxi - float(f.subs(x,x2))
        matrix.append([3,x2,fxi,error])
        det = (x2-x1)-(x1-x0)
        counter = 4
        xi = 0
        while(fxi != 0 and error > tol and counter < n and det!=0):
            xi = x2 - ( (x2 - x1)**2 )/det
            fxi = float(f.subs(x,xi))
            det = (x2-x1)-(x1-x0)
            error = abs(xi-x2)
            matrix.append([counter,xi,fxi,error])
            x0 = xi
            x1 = float(f.subs(x,x0))
            x2 = float(f.subs(x,x1))
            counter += 1
        if fxi == 0:
            ans = 'The root has been found and it is: ', xi
        elif error <= tol:
            ans = 'An approximation has been found and is: ', xi
        elif det == 0:
            ans = 'Error during method execution'
        else:
            ans = 'The method fails with the maximum number of iterations given'
        return ans,matrix

# f = sm.ln(sm.sin(x) ** 2+1)-(1/2)
# x0 = 1
# tol = 1e-7
# niter = 20

# ans,matrix = aitken(f,x0,tol,niter)
# print(ans)
# print(matrix)
# print(ans)
# print()
# if len(matrix)>1:
#   print(f"%11s | %15s | %15s | %15s"%(matrix[0][0],matrix[0][1],matrix[0][2],matrix[0][3]))
#   i = 1
#   while (i < len(matrix)):
#     print(f"%11s | %15E | %15E | %15E"%(matrix[i][0],matrix[i][1],matrix[i][2],matrix[i][3]))
#     i +=1